lib/plsql/procedure.rb

```diff
@@ -350,18 +371,35 @@ def get_field_definitions(argument_metadata) #:nodoc:
       case argument_metadata[:type_object_type]
       when "PACKAGE"
         @schema.select_all(
-          "SELECT attr_no, attr_name, attr_type_owner, attr_type_name, attr_type_package, length, precision, scale, char_used
-          FROM ALL_PLSQL_TYPES t, ALL_PLSQL_TYPE_ATTRS ta
-          WHERE t.OWNER = :owner AND t.type_name = :type_name AND t.package_name = :package_name
-          AND ta.OWNER = t.owner AND ta.TYPE_NAME = t.TYPE_NAME AND ta.PACKAGE_NAME = t.PACKAGE_NAME
-          ORDER BY attr_no",
-          @schema_name, argument_metadata[:type_name], argument_metadata[:type_subname]) do |r|
+          "SELECT ta.attr_no, ta.attr_name,
+            nvl(s.table_owner, attr_type_owner) attr_type_owner,
+            CASE WHEN ta.attr_type_package IS NOT NULL THEN ta.attr_type_name ELSE nvl(s.table_name, ta.attr_type_name) END attr_type_name,
+            CASE WHEN ta.attr_type_package IS NOT NULL THEN nvl(s.table_name, ta.attr_type_package) END attr_type_package,
+            ta.length, ta.precision, ta.scale, ta.char_used
+          FROM all_plsql_type_attrs ta
+          LEFT JOIN all_synonyms s ON ta.attr_type_owner = s.owner AND nvl(ta.attr_type_package, ta.attr_type_name) = s.synonym_name
+          WHERE ta.owner = :owner
+            AND ta.type_name = :type_name
+            AND ta.package_name = :package_name
+          ORDER BY ta.attr_no",
+          argument_metadata[:type_owner], argument_metadata[:type_name], argument_metadata[:type_subname]) do |r|
 
           attr_no, attr_name, attr_type_owner, attr_type_name, attr_type_package, attr_length, attr_precision, attr_scale, attr_char_used = r
 
+          attr_type_name = 'PLS_INTEGER' if attr_type_name == 'PL/SQL PLS INTEGER'
+          attr_type_name = 'BINARY_INTEGER' if attr_type_name == 'PL/SQL BINARY INTEGER'
+
+          composite_type = nil
+          if attr_type_owner != nil
+            if attr_type_package != nil
+              composite_type = get_composite_type(attr_type_owner, attr_type_name, attr_type_package)
+            else
```

Comment:
Hi @joschug,
If I read the changes correctly, I see that we modified "get_composite_type" to handle a few failures. For example, when the "attr_type_package" is nil for some datatypes and then it throws "bind type is not given." (at least for me I get this error). So, with this PR, we are defaulting to "composite_type = 'TABLE'" when attr_type_package is nil. Please correct me if my understanding is wrong.

In my codebase, I'm using oracle's "SYS.XMLTYPE" datatype as one of the elements in my user-defined type. For this case also, composite_type is set to TABLE which I guess is not right. I'm not an expert in ruby and somehow managed to read this PR. Can you please take a look at whether this PR support XMLTYPE and other generic datatypes in Oracle?

Thanks,